net/sched/act_skbmod.c

```diff
@@ -32,15 +32,30 @@ static int tcf_skbmod_run(struct sk_buff *skb, const struct tc_action *a,
 	struct tcf_skbmod_params *p;
 	u64 flags;
 	int err;
+	int ip_len;
 
 	tcf_lastuse_update(&d->tcf_tm);
 	bstats_cpu_update(this_cpu_ptr(d->common.cpu_bstats), skb);
 
+	switch (tc_skb_protocol(skb)) {
+	case htons(ETH_P_IP):
+		ip_len = sizeof(struct iphdr);
+		break;
+
+	case htons(ETH_P_IPV6):
+		ip_len = sizeof(struct ipv6hdr);
+		break;
+
+	default:
+		ip_len = 0;
+		break;
+	}
+
 	/* XXX: if you are going to edit more fields beyond ethernet header
 	 * (example when you add IP header replacement or vlan swap)
 	 * then MAX_EDIT_LEN needs to change appropriately
 	*/
-	err = skb_ensure_writable(skb, MAX_EDIT_LEN);
+	err = skb_ensure_writable(skb, MAX_EDIT_LEN + ip_len);
```

Comment:
We only **read** the DS field, so we don't need to check that the IP header is writable, only that there's an IP header.